caron.js

```diff
@@ -2,270 +2,253 @@
+      if (this.status.processing) return
+      if (this.debug) console.log('stopped')
+
+      clearInterval(this.stopInter)
+      cb(null)
+    }, 250)
   }
 
-  const args = []
-  var startTs = null
+  start () {
+    this.work()
 
-  if (debug) {
-    startTs = process.hrtime()
+    return this
   }
 
-  args.push(
-    ptype,
-    Date.now(), // timestamp
-    rseed
-  )
+  kill () {
+    if (!this.kill) return
+    process.exit()
+  }
 
-  args.push(
-    (err, res) => {
-      STATUS.processing = 0
+  work () {
+    const debug = this.debug
 
-      if (err) {
-        console.error(err)
-        stop()
-        return
-      }
+    if (!this.status.active || this.status.processing) return
 
-      if (debug) {
-        let elapsed = elapsedTime(startTs)
-        console.log(res[1] + ' jobs processed in ' + elapsed[0] + 's,' + Math.round(elapsed[1] / 1000) + 'µs')
-      }
+    const rseed = Date.now()
 
-      setTimeout(work, program.freq || 25)
+    if (rseed < this.status.rseed) {
+      setTimeout(() => {
+        this.work()
+      }, 5)
+      return
     }
-  )
 
-  STATUS.processing = 1
+    const args = []
+    let startTs = null
 
-  redis.qwork.apply(
-    redis, args
-  )
-}
+    if (debug) {
+      startTs = process.hrtime()
+    }
 
-work()
+    args.push(
+      this.ptype,
+      Date.now(), // timestamp
+      rseed
+    )
+
+    args.push(
+      (err, res) => {
+        this.status.processing = 0
+
+        if (err) {
+          console.error(err)
+          this.stop(() => {
+            this.kill()
+          })
+          return
+        }
+
+        if (debug) {
+          let elapsed = elapsedTime(startTs)
+          console.log(res[1] + ' jobs processed in ' + elapsed[0] + 's,' + Math.round(elapsed[1] / 1000) + 'µs')
+        }
+
+        setTimeout(() => {
+          this.work()
+        }, this.freq || 25)
+      }
+    )
 
-const stop = () => {
-  if (!STATUS.active) return
+    this.status.processing = 1
 
-  STATUS.active = 0
+    this.redis.qwork.apply(
+      this.redis, args
+    )
+  }
 
-  if (debug) console.log('stopping...')
+  registerHandlers () {
+    Redis.Promise.onPossiblyUnhandledRejection(e => {
+      console.error(e)
+      this.status.processing = 0
+      this.stop(() => {
+        this.kill()
+      })
+    })
+
+    this.redis.on('error', e => {
+      this.status.processing = 0
+      console.log(e)
+    })
+  }
 
-  setInterval(() => {
-    if (STATUS.processing) return
-    if (debug) console.log('stopped')
-    process.exit()
-  }, 250)
+  setupRedis (scripts) {
+    const debug = this.debug
+
+    const script = [
+      scripts.prefix,
+      scripts[this.ptype].lua,
+      scripts.suffix
+    ].join('\n')
+
+    if (debug) console.log(script)
+
+    this.redis.defineCommand('qwork', {
+      lua: script,
+      numberOfKeys: 0
+    })
+  }
+
+  setupScripts (program) {
+    const scripts = {
+      prefix: [
+        'math.randomseed(tonumber(ARGV[3]))',
```

Comment:
Block strings would look great here :D Or breaking the lua out to its own file to get syntax highlighting.

Reply:
aw right, maybe we can get this merged first - in the next pr i would put it into a seperate file.